internal/generate/generate.go

```diff
@@ -0,0 +1,267 @@
+	"github.com/cloudflare/cfssl/signer/local"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "generate")
+)
+
+// CA contains the data and metadata of the certificate and keyfile
+type Cert struct {
+	CommonName         string
+	ValidityDuration   time.Duration
+	Usage              []string
+	K8sSecretName      string
+	K8sSecretNamespace string
+
+	CertBytes []byte
+	KeyBytes  []byte
+}
+
+// NewCert creates a new certificate blueprint
+func NewCert(
+	commonName string,
+	validityDuration time.Duration,
+	usage []string,
+	k8sSecretName string,
+	k8sSecretNamespace string,
+) *Cert {
+	return &Cert{
+		CommonName:         commonName,
+		ValidityDuration:   validityDuration,
+		Usage:              usage,
+		K8sSecretName:      k8sSecretName,
+		K8sSecretNamespace: k8sSecretNamespace,
+	}
+}
+
+// Generate the certificate and keyfile and populate c.CertBytes and c.CertKey
+func (c *Cert) Generate(ca *x509.Certificate, caSigner crypto.Signer) error {
+	log.WithFields(logrus.Fields{
+		logfields.CertCommonName:       c.CommonName,
+		logfields.CertValidityDuration: c.ValidityDuration,
+		logfields.CertUsage:            c.Usage,
+	}).Info("Generating certificate")
+
+	certRequest := &csr.CertificateRequest{
+		CN:         c.CommonName,
+		Hosts:      []string{c.CommonName},
+		KeyRequest: csr.NewKeyRequest(),
+	}
+
+	g := &csr.Generator{Validator: genkey.Validator}
+	csrBytes, keyBytes, err := g.ProcessRequest(certRequest)
+	if err != nil {
+		return err
+	}
+
+	policy := &config.Signing{
+		Default: &config.SigningProfile{
+			Usage:  c.Usage,
+			Expiry: c.ValidityDuration,
+		},
+	}
+	s, err := local.NewSigner(caSigner, ca, signer.DefaultSigAlgo(caSigner), policy)
+	if err != nil {
+		return err
+	}
+
+	signReq := signer.SignRequest{Request: string(csrBytes)}
+	certBytes, err := s.Sign(signReq)
+	if err != nil {
+		return err
+	}
+
+	c.CertBytes = certBytes
+	c.KeyBytes = keyBytes
+	return nil
+}
+
+// StoreAsSecret creates or updates the certificate and keyfile in a K8s secret
+func (c *Cert) StoreAsSecret(k8sClient *kubernetes.Clientset) error {
+	if c.CertBytes == nil || c.KeyBytes == nil {
+		return fmt.Errorf("cannot create secret %s/%s from empty certificate",
+			c.K8sSecretNamespace, c.K8sSecretName)
+	}
+
+	log.WithFields(logrus.Fields{
+		logfields.K8sSecretNamespace: c.K8sSecretNamespace,
+		logfields.K8sSecretName:      c.K8sSecretName,
+	}).Info("Creating K8s Secret")
+
+	secret := &v1.Secret{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:      c.K8sSecretName,
+			Namespace: c.K8sSecretNamespace,
+		},
+		Data: map[string][]byte{
+			"tls.crt": c.CertBytes,
+			"tls.key": c.KeyBytes,
+		},
+		Type: v1.SecretTypeTLS,
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.K8sRequestTimeout)
```

Comment:
nit: for future-proofing, we should probably change the signature of `StoreAsSecret` to take `ctx context.Context` as a parameter.

Reply:
Good idea. It also removes the dependency on the `option` package, which was unintended actually.